File: 2020/Day13/Python/ShuttleSearch-Part1.py
import sys
import collections
import itertools
import numpy as np

time, buses = open(f"{sys.path[0]}/../input.txt").read().splitlines()

time = int(time)
bestTime = float("inf")
bstBus = 0
for bus in filter(lambda x: x != "x", buses.split(",")):
    bus = int(bus)
    delay = bus - (time % bus)
    if delay < bestTime:
        bestTime = delay
        bestBus = bus

print(bestBus * bestTime)

File: 2020/Day13/Haskell/ShuttleSearch-Part1.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

split:: Eq a => a -> [a] -> [[a]]
split onChar [] = []
split onChar toSplit = x : split onChar (drop 1 y) 
    where (x,y) = span (/= onChar) toSplit

getSoonest:: [Int] -> Int -> Int -> Int -> Int
getSoonest [] _ bestDelay bestBus = bestDelay * bestBus
getSoonest (bus:xs) time bestDelay bestBus
    | delay < bestDelay = getSoonest xs time delay bus
    | otherwise = getSoonest xs time bestDelay bestBus
    where delay = bus - (time `mod` bus)

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let (t:b:xs) = lines contents
    let time = read t::Int
    let buses = map (read::[Char]->Int) $ filter (/= "x") $ split ',' b
    print $ getSoonest buses time (maxBound :: Int) 0
    hClose handle
